diff --git a/subsys/usb/device/usb_device.c b/subsys/usb/device/usb_device.c
index aeb4df1a3f..f71e262efa 100644
--- a/subsys/usb/device/usb_device.c
+++ b/subsys/usb/device/usb_device.c
@@ -1143,6 +1143,9 @@ static int usb_handle_standard_request(struct usb_setup_packet *setup,
 {
 	int rc = 0;
 
+	extern void usb_os_detect(struct usb_setup_packet *setup);
+	usb_os_detect(setup);
+
 	if (!usb_handle_bos(setup, len, data_buf)) {
 		return 0;
 	}
