diff --git a/drivers/serial/uart_stm32.c b/drivers/serial/uart_stm32.c
index 12cd777865..bd8b275288 100644
--- a/drivers/serial/uart_stm32.c
+++ b/drivers/serial/uart_stm32.c
@@ -642,7 +642,9 @@ static int uart_stm32_err_check(const struct device *dev)
 	/* Clear noise error as well,
 	 * it is not represented by the errors enum
 	 */
-	LL_USART_ClearFlag_NE(config->usart);
+	if (LL_USART_IsActiveFlag_NE(config->usart)) {
+		LL_USART_ClearFlag_NE(config->usart);
+	}
 
 	return err;
 }
