diff --git a/drivers/serial/uart_mcux.c b/drivers/serial/uart_mcux.c
index 9e894addb6..ffefb082ff 100644
--- a/drivers/serial/uart_mcux.c
+++ b/drivers/serial/uart_mcux.c
@@ -164,9 +164,11 @@ static int uart_mcux_err_check(const struct device *dev)
 		err |= UART_ERROR_FRAMING;
 	}
 
-	UART_ClearStatusFlags(config->base, kUART_RxOverrunFlag |
-					    kUART_ParityErrorFlag |
-					    kUART_FramingErrorFlag);
+	if (err) {
+		UART_ClearStatusFlags(config->base, kUART_RxOverrunFlag |
+											kUART_ParityErrorFlag |
+											kUART_FramingErrorFlag);
+	}
 
 	return err;
 }
diff --git a/soc/arm/nxp_kinetis/k2x/soc.c b/soc/arm/nxp_kinetis/k2x/soc.c
index 67ea118340..d28d79394d 100644
--- a/soc/arm/nxp_kinetis/k2x/soc.c
+++ b/soc/arm/nxp_kinetis/k2x/soc.c
@@ -118,6 +118,9 @@ static ALWAYS_INLINE void clock_init(void)
 
 static int fsl_frdm_k22f_init(void)
 {
+	/* disable MPU (using MPU_CESR) */
+	*((uint32_t*) 0x4000D000) = 0;
+
 	/* release I/O power hold to allow normal run state */
 	PMC->REGSC |= PMC_REGSC_ACKISO_MASK;
 
