diff --git a/soc/arm/st_stm32/stm32g0/soc.c b/soc/arm/st_stm32/stm32g0/soc.c
index cf8f4fb937..5179db30ce 100644
--- a/soc/arm/st_stm32/stm32g0/soc.c
+++ b/soc/arm/st_stm32/stm32g0/soc.c
@@ -99,7 +99,7 @@ static int stm32g0_init(const struct device *arg)
 	SystemCoreClock = 16000000;
 
 	/* Disable the internal Pull-Up in Dead Battery pins of UCPD peripheral */
-	stm32g0_disable_dead_battery();
+	// stm32g0_disable_dead_battery();
 
 	return 0;
 }
