diff --git a/drivers/usb/device/usb_dc_mcux.c b/drivers/usb/device/usb_dc_mcux.c
index 0667486b62..ad6d2ec91d 100644
--- a/drivers/usb/device/usb_dc_mcux.c
+++ b/drivers/usb/device/usb_dc_mcux.c
@@ -95,6 +95,8 @@ BUILD_ASSERT(NUM_INSTS <= 1, "Only one USB device supported");
 #elif DT_NODE_HAS_STATUS(DT_NODELABEL(usb2), okay)
 #define CONTROLLER_ID kUSB_ControllerEhci1
 #endif /* IMX RT */
+#elif defined(CONFIG_SOC_SERIES_KINETIS_K6X)
+#define CONTROLLER_ID kUSB_ControllerEhci0
 #else
 /* If SOC has EHCI or LPCIP3511 then probably just need to add controller ID to this code */
 #error "USB driver does not yet support this SOC"
